### ISSUE TITLE
Shamir

### ISSUE THREAD

**Issue by cryptoida** [stepansnigirev/javacard_playground#1]
Hi Stepan,

I would like to know if the Shamir Secret Sharing can be implement in javacard. ( as a applet)
The idea would be to store secure mnemonic seed to restore it later, if necessary.

Shamirs secret sharing scheme, can be used,  to split a seed mnemonic for crypto wallets to multiple mnemonics and securely store each split in a distinct javacard.

Ex : 
Split Parameters :

- allow reconstruction with  3 of 5 parts. (need 5 javacards)
- combine minumum 3 shares (3 javacards), to restore original seed mnemonic.

Sources :

SLIP-0039 : Shamir's Secret-Sharing for Mnemonic Codes
https://github.com/satoshilabs/slips/blob/master/slip-0039.md#IndexEncoding

https://iancoleman.io/shamir39/
https://iancoleman.io/shamir/
https://github.com/Qbicz/multi-secret-sharing

Please let me know.

**Comment by stepansnigirev** [comment]
Yes, I don't see why not.
For Shamir sharing implementation on the card one would need modulo multiplication and inversion. These operations can be implemented using RSA engine present on modern cards.

How do you imagine that hardware-wise? N+1 javacards where N are shamir shares and 1 is responsible for reconstruction of the secret and signing? Or what kind of device will be doing the splitting and reconstruction of the secret?

Until we have Schnorr in Bitcoin we need to reconstruct the full secret somewhere every time we sign. When Schnorr is deployed we can avoid reconstruction the full secret and gather partial signatures from the shares - here your shamir-javacard idea will really shine.

**Comment by stepansnigirev** [comment]
Ok, I looked more carefully through the specs of slip-39, one thing that makes it difficult is the encryption of the master secret using PBKDF2 with 2500 iterations.
https://github.com/satoshilabs/slips/blob/master/slip-0039.md#encryption-of-the-master-secret

Recently I checked how fast is `PBKDF2` on the card and it was terribly slow - around a minute with 2048 iterations (I was testing BIP-39). This might be a problem for implementation or limit you to a small subset of the cards that have decent hardware accelerator with support of `ALG_HMAC_SHA512`.
